File: code/java/pq/src/main/java/org/swufe/datastructures/MaxPQ.java
package org.swufe.datastructures;

import java.util.Arrays;
import java.util.NoSuchElementException;

public class MaxPQ<Key extends Comparable<Key>> {

    private Key[] pq;
    private int n;

    @SuppressWarnings("unchecked")
    public MaxPQ() {
        int INIT_CAPACITY = 16;
        pq = (Key[]) new Comparable[INIT_CAPACITY + 1];
        n = 0;
    }

    public int size() {
        return n;
    }

    public boolean isEmpty() {
        return n == 0;
    }

    private void swap(int i, int j) {
        Key t = pq[i];
        pq[i] = pq[j];
        pq[j] = t;
    }

    private boolean less(int i, int j) {
        return pq[i].compareTo(pq[j]) < 0;
    }

    private void swim(int k) {
        while (k > 1 && less(k/2, k)) {
            swap(k/2, k);
            k /= 2;
        }
    }

    private void sink(int k) {
        while (2 * k <= n) {
            int j = 2 * k;
            if (j < n && less(j, j+1)) j++;
            if (!less(k, j)) break;
            swap(k, j);
            k = j;
        }
    }

    private void resize(int capacity) {
        assert capacity > n;
        pq = Arrays.copyOf(pq, capacity);
    }

    public void insert(Key key) {
        if (n == pq.length - 1) resize(2 * pq.length);
        pq[++n] = key;
        swim(n);
        assert isMaxHeap();
    }

    public Key max() {
        if (isEmpty()) throw new NoSuchElementException();
        return pq[1];
    }

    public Key delMax() {
        if (isEmpty()) throw new NoSuchElementException();
        Key root = pq[1];
        swap(1, n--);
        sink(1);
        pq[n+1] = null;
        if (n > 0 && n == (pq.length - 1) / 4) resize((pq.length - 1) / 2);
        assert isMaxHeap();
        return root;
    }

    // is the heap rooted at k a max heap?
    private boolean isMaxHeapOrdered(int k) {
        if (k > n) return true;
        int left = 2 * k;
        int right = left + 1;
        if (left <= n && less(k, left)) return false;
        if (right <= n && less(k, right)) return false;
        return isMaxHeapOrdered(left) && isMaxHeapOrdered(right);
    }

    private boolean isMaxHeap() {
        // 1..=n is not null
        for (int i = 1; i <= n; i++) {
            if (pq[i] == null) return false;
        }
        // n+1..length is null
        for (int i = n + 1; i < pq.length; i++) {
            if (pq[i] != null) return false;
        }
        if (pq[0] != null) return false;
        return isMaxHeapOrdered(1);
    }
}

File: code/python/pq/max_pq.py
from no_element import NoElement


class MaxPQ:
    """A max binary heap."""

    def __init__(self, data=None):
        self._pq = []
        self._pq.append(None)  # append a dummy key at index 0
        if data is not None:
            self._pq.extend(data)
            n = len(data)
            for k in range(n // 2, 0, -1):
                self._sink(k)
            assert self._is_max_heap()

    def size(self):
        return len(self._pq) - 1

    def is_empty(self):
        return self.size() == 0

    def _swap(self, i, j):
        self._pq[i], self._pq[j] = self._pq[j], self._pq[i]

    def _less(self, i, j):
        return self._pq[i] < self._pq[j]

    def _swim(self, k):
        while k > 1 and self._less(k//2, k):
            self._swap(k//2, k)
            k //= 2

    def _sink(self, k):
        while 2 * k <= self.size():
            j = 2 * k
            if j < self.size() and self._less(j, j + 1):
                j += 1
            if not self._less(k, j):
                break
            self._swap(k, j)
            k = j

    def insert(self, key):
        self._pq.append(key)
        self._swim(self.size())
        assert self._is_max_heap()

    def max(self):
        if self.is_empty():
            raise NoElement
        return self._pq[1]

    def del_max(self):
        if self.is_empty():
            raise NoElement
        root = self._pq[1]
        self._swap(1, self.size())
        self._pq.pop()
        self._sink(1)
        assert self._is_max_heap()
        return root

    def _is_max_heap_ordered(self, k):
        if k > self.size():
            return True
        left = 2 * k
        right = 2 * k + 1
        if left <= self.size() and self._less(k, left):
            return False
        if right <= self.size() and self._less(k, right):
            return False
        return self._is_max_heap_ordered(left) and self._is_max_heap_ordered(right)

    def _is_max_heap(self):
        if self._pq[0] is not None:
            return False
        return self._is_max_heap_ordered(1)
